river/command/map.zig

```diff
@@ -53,35 +53,41 @@ pub fn map(
     if (args.len < 5) return Error.NotEnoughArguments;
 
     const mode_id = try modeNameToId(allocator, seat, args[1], out);
-    const modifiers = try parseModifiers(allocator, args[2], out);
+    const optionals = parseOptionalArgs(args[2..]);
+    // add argument count so that args[i] will always be set to the modifiers part of args
+    const i = optionals.i + 2;
+    const modifiers = try parseModifiers(allocator, args[i], out);
 
     // Parse the keysym
-    const keysym_name = try std.cstr.addNullByte(allocator, args[3]);
+    const keysym_name = try std.cstr.addNullByte(allocator, args[i + 1]);
     defer allocator.free(keysym_name);
     const keysym = c.xkb_keysym_from_name(keysym_name, .XKB_KEYSYM_CASE_INSENSITIVE);
     if (keysym == c.XKB_KEY_NoSymbol) {
         out.* = try std.fmt.allocPrint(
             allocator,
             "invalid keysym '{}'",
-            .{args[3]},
+            .{args[i + 1]},
         );
         return Error.Other;
     }
 
+    const mapping = try Mapping.init(util.gpa, keysym, modifiers, optionals.release, args[4..]);
+    errdefer mapping.deinit(util.gpa);
+
     // Check if the mapping already exists
     const mode_mappings = &seat.input_manager.server.config.modes.items[mode_id].mappings;
-    for (mode_mappings.items) |existant_mapping| {
-        if (existant_mapping.modifiers == modifiers and existant_mapping.keysym == keysym) {
+    for (mode_mappings.items) |*existant_mapping| {
+        if (mapping.isEqual(existant_mapping)) {
```

Comment:
Why not allow bindings for both press and release of a given key?

Reply:
Oh, I missed adding the check for release.